logicalplan/plan.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"time"
+
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+var DefaultOptimizers = []Optimizer{
+	SortMatchers{},
+	MergeSelectsOptimizer{},
+}
+
+type Plan interface {
+	RunOptimizers([]Optimizer) parser.Expr
+}
+
+type Optimizer interface {
```

Comment:
Maybe renaming to `RuleBasedOptimizer` would be good here? In case we choose to implement plan cost estimation in the future? 🙂 
```suggestion
type RuleBasedOptimizer interface {
```

Reply:
Let's maybe introduce such distinctions when we have different optimizer types?

Reply:
Sure!

Reply:
👍🏽